packages/abi/src/parser/specifications/v1/resolved-type.ts

```diff
@@ -0,0 +1,42 @@
+import type { AbiType } from '../../abi';
+
+export interface ResolvedComponent {
+  name: string;
+  type: ResolvedType;
+}
+
+export class ResolvedType {
+  public type: string;
+  public typeId: string | number;
+  public components: ResolvedComponent[] | undefined;
+  public typeParamsArgsMap: Array<[number, ResolvedType]> | undefined;
+  private metadataTypeId: number | undefined;
+
+  constructor(params: {
+    type: string;
+    typeId: string | number;
+    components?: ResolvedComponent[];
+    typeParamsArgsMap?: Array<[number, ResolvedType]>;
+    metadataTypeId?: number;
+  }) {
+    this.type = params.type;
+    this.typeId = params.typeId;
+    this.components = params.components;
+    this.typeParamsArgsMap = params.typeParamsArgsMap;
+    this.metadataTypeId = params.metadataTypeId;
+  }
+
+  public toAbiType(): AbiType {
+    return {
+      swayType: this.type,
+      concreteTypeId: this.typeId as string,
```

Comment:
Further details after testing.

Reply:
@petertonysmith94 I'm not sure what you mean,  is it that there are `concreteTypeId`s  that can be `metadataTypeId`s as a result of testing?

Reply:
The test in [`abi-parser.test.ts`](https://github.com/FuelLabs/fuels-ts/pull/3089/files#diff-bb55bd22085ec8047971b553940d03af4e2cd947a627b18dc5dd7111410d6d07) now ensures that this cast `as string` is always correct and the `concreteTypeId` can never be a `metadataTypeId`. The `typeId` field can still be a `number` because of the `toComponentType` method above, which covers both concrete and metadata components.